### ISSUE TITLE
[CGNS-146] Can't compile using MSVC & IVF on windows (MSVS 14)

### ISSUE THREAD

**Issue by brtnfld** [CGNS/CGNS#542]
> This issue has been imported from JIRA. Read the [original ticket here](https://cgnsorg.atlassian.net/browse/CGNS-146).

- _**Created at:**_ Sat, 15 Sep 2018 15:24:01 -0500
- _**Resolved at:**_ Mon, 4 Mar 2019 10:41:29 -0600

<p>A requirement of a project I am working on is to statically link against CGNS on windows, including the Fortran interfaces.</p>

<p>I have tried to compile on Windows 10, using the CMake files and using the "Visual Studio 14 2015 Win64" generator. The C compiler is the default MSVC compiler, and the fortran compiler is Intel IVF 18.0.3.</p>

<p>I have tried setting the CMake configuration according to both the "Building on Windows" section of the FAQ, and the configure options being passed to CMake in the AppVeyor CI testing in the main CGNS repository.</p>

<p>I have downloaded the v3.3.1 release from the github releases page.</p>

<p><b>UPDATE: A work around has been identified</b></p>

<p>The work around is to explicitly set the linker language to C for `cgns_static`</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-none">
set_property(TARGET cgns_static PROPERTY LINKER_LANGUAGE C)
</pre>
</div></div>

**Comment by brtnfld** [comment]
> This comment has been imported from JIRA. Read the [original comment here](https://cgnsorg.atlassian.net/browse/CGNS-146#comment-13824).

- _**Created at:**_ Thu, 27 Sep 2018 20:08:40 -0500

<p>How do I get someone to respond to this issue? Simply enabling Fortran on a windows build causes a problem. (At least when you are trying to use only static libraries.)</p>

<p>I've done some more debugging, and by simply changing Fortran from OFF to ON, the cgns.lib is no longer produced by MSVS during compilation, even though it claims that the cgns_static target/project is up to date. Turning Fortran OFF causes it to actually produce the cgns.lib file.</p>

<p>I have been digging into your build system, but I still don't understand why this is happening. It seems as though there are some bad side effects from some CMake commands. I am using CMake 3.11.</p>

**Comment by brtnfld** [comment]
> This comment has been imported from JIRA. Read the [original comment here](https://cgnsorg.atlassian.net/browse/CGNS-146#comment-13833).

- _**Created at:**_ Tue, 4 Dec 2018 09:42:55 -0600

<p>Update: Brad King from Kitware found the issue with building on windows with Fortran enabled and newer CMake. Here is what he says, and what the work around is. The problem is that CMake is not setting the linker language correctly when adding a Fortran object library as a C library source:</p>

<blockquote>
<p>I can reproduce this using a tiny project:</p>

<p>```<br/>
cmake_minimum_required(VERSION 3.8)<br/>
project(Mixed C Fortran)</p>

<p>add_library(fobj OBJECT f.f)<br/>
add_library(mixed c.c $<TARGET_OBJECTS:fobj>)<br/>
```</p>

<p>The cause is the changes discussed here:</p>

<ul>
	<li><a href="https://gitlab.kitware.com/cmake/cmake/issues/16738" class="external-link" rel="nofollow noreferrer">https://gitlab.kitware.com/cmake/cmake/issues/16738</a></li>
	<li><a href="https://gitlab.kitware.com/cmake/cmake/merge_requests/643" class="external-link" rel="nofollow noreferrer">https://gitlab.kitware.com/cmake/cmake/merge_requests/643</a></li>
</ul>


<p>A workaround is:</p>

<p>```<br/>
set_property(TARGET mixed PROPERTY LINKER_LANGUAGE C)<br/>
```</p>

<p>Replace "mixed" with "cgns_static" in CGNS.</p></blockquote>

**Comment by brtnfld** [comment]
> This comment has been imported from JIRA. Read the [original comment here](https://cgnsorg.atlassian.net/browse/CGNS-146#comment-13834).

- _**Created at:**_ Tue, 4 Dec 2018 10:06:02 -0600

<p>PR opened at <a href="https://github.com/CGNS/CGNS/pull/87" class="external-link" rel="nofollow noreferrer">https://github.com/CGNS/CGNS/pull/87</a>. Testing locally on Windows to confirm fix.</p>

**Comment by brtnfld** [comment]
> This comment has been imported from JIRA. Read the [original comment here](https://cgnsorg.atlassian.net/browse/CGNS-146#comment-13835).

- _**Created at:**_ Wed, 5 Dec 2018 09:11:01 -0600

<p>I can confirm that the build now succeeds on Windows under MSVS with Fortran enabled.</p>

<p>A corresponding issue has been opened on teh CMake bug tracker: <a href="https://gitlab.kitware.com/cmake/cmake/issues/18687" class="external-link" rel="nofollow noreferrer">https://gitlab.kitware.com/cmake/cmake/issues/18687</a></p>

<p>A minor problem remains: The installation of the cgns `.mod` file fails due to the way MSVS behaves. The CMake installation is not looking under the `$<CONFIG>` subdirectory for the module file, and thus the installation of the .mod file is failing. I'll try to add a fix for that here as well.</p>

**Comment by brtnfld** [comment]
> This comment has been imported from JIRA. Read the [original comment here](https://cgnsorg.atlassian.net/browse/CGNS-146#comment-13836).

- _**Created at:**_ Fri, 14 Dec 2018 11:42:45 -0600

<p>PR: <a href="https://github.com/CGNS/CGNS/pull/87" class="external-link" rel="nofollow noreferrer">https://github.com/CGNS/CGNS/pull/87</a> is ready to merge, AFAICT. Please let me know if anyone has any concerns that I can address, changes they would like to see, etc.</p>

**Comment by brtnfld** [comment]
> This comment has been imported from JIRA. Read the [original comment here](https://cgnsorg.atlassian.net/browse/CGNS-146#comment-13837).

- _**Created at:**_ Mon, 11 Feb 2019 12:48:27 -0600

<p>Can this be closed?</p>

**Comment by brtnfld** [comment]
> This comment has been imported from JIRA. Read the [original comment here](https://cgnsorg.atlassian.net/browse/CGNS-146#comment-13838).

- _**Created at:**_ Mon, 11 Feb 2019 12:56:52 -0600

<p>I think so. I haven't had time to test the latest release. I've also noticed that MSVS does not seem to do well with mixed language targets. I have doubts that the object library is linking correctly when built with MSVS. But for not, this seems resolved enough.</p>

<p>For the application/project I am working on, in the end, my test seemed to fail due to HDF5, which is a different kettle of fish. It is possible that bugs remain with Intel Fortran and MSVS, but we've factored out the code requiring CGNS, and added flags to turn it on and off. So for now, we are skipping CGNS on Windows.</p>

<p>I would recommend, even though it's obnoxious, compiling fortran and C libraries separately to better accommodate MSVS. This is somewhat different than the present issues, so I am happy to mark this as resolved, at least for now.</p>
